doc/api/crypto.md

```diff
@@ -1813,28 +1813,35 @@ must be an object with the properties described above.
 <!-- YAML
 added: v11.6.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/26278
+    description: The `key` argument can now be a `KeyObject` with type
+                 `private`.
   - version: v11.7.0
     pr-url: https://github.com/nodejs/node/pull/25217
     description: The `key` argument can now be a private key.
 -->
-* `key` {Object | string | Buffer}
+* `key` {Object | string | Buffer | KeyObject}
   - `key`: {string | Buffer}
   - `format`: {string} Must be `'pem'` or `'der'`. **Default:** `'pem'`.
   - `type`: {string} Must be `'pkcs1'` or `'spki'`. This option is required
     only if the `format` is `'der'`.
 * Returns: {KeyObject}
 
 Creates and returns a new key object containing a public key. If `key` is a
-string or `Buffer`, `format` is assumed to be `'pem'`; otherwise, `key`
-must be an object with the properties described above.
+string or `Buffer`, `format` is assumed to be `'pem'`; if `key` is a `KeyObject`
+with type `'private'`, the public key is derived from the given private key;
+otherwise, `key` must be an object with the properties described above.
 
 If the format is `'pem'`, the `'key'` may also be an X.509 certificate.
 
 Because public keys can be derived from private keys, a private key may be
 passed instead of a public key. In that case, this function behaves as if
 [`crypto.createPrivateKey()`][] had been called, except that the type of the
-returned `KeyObject` will be `public` and that the private key cannot be
-extracted from the returned `KeyObject`.
+returned `KeyObject` will be `'public'` and that the private key cannot be
+extracted from the returned `KeyObject`. Similarly, if a `KeyObject` with type
```

Comment:
Is there a cost related to this action? When we pass a `'private'` key to functions needing a `'public'` one, is the key being derived over and over?

Should an API be encouraged to prepare and cache the public KeyObject and pass that one instead?

Reply:
The only cost is the creation of two or three JavaScript objects. In cryptography, we often speak about "deriving the key", but in this case, that does not actually happen. For all supported key types (RSA, DSA, EC), the public key is a part of the private key, so OpenSSL just keeps a memory reference to the private key instead of actually creating a public key. Node.js makes sure that only the public portion of the referenced key can be accessed.

> Should an API be encouraged to prepare and cache the public KeyObject and pass that one instead?

I don't think calling `createPublicKey(privateKeyObject)` takes a considerable amount of time, but I can try to benchmark it later. In general, key objects should be cached, but in this case, it should be sufficient to cache the private key object. (You can also pass the private key object instead of a public key object to all APIs that take public keys, so there is usually little need to explicitely derive a public key object.) Is that okay? :)

Reply:
Yes, absolutely. I assumed this to be the case, just wanted to make sure. I'm making use of KeyObjects only in my [JOSE module](https://github.com/panva/jose/) and when i've got private one at hand i pass that one to the crypto operation. I was wondering if there would be any reason to do `createPublic(private)` before calling crypto instead, but since it's essentially just "cherry-picking" what it needs to from an already prepared openssl key representation that's not needed. Thank you for confirming.

Reply:
Exactly, there is no performance gain in calling `createPublic(private)`! Thanks for the excellent question!